### PR TITLE
Add "state save" and "state load" commands

### PR DESCRIPTION
**Pull request by j-jzk**
This PR adds two new commands:
* `haxelib state save haxelib-lock.hxml`
* `haxelib state load haxelib-lock.hxml`

This manifest file (note that `.hxml` extension isn't mandatory here) is used to "lock" the current state of the libraries in the project. This is useful for reproducing the exact state of dependencies in a project (on a new machine or in Docker), especially when working with libraries installed from git.

The intended workflow is following:
1. do package management operations
2. lock the dependencies: `haxelib state save haxelib-lock.hxml`
3. apply the manifest file in another environment: `haxelib state load haxelib-lock.hxml`

This manifest mechanism supports published haxelib versions, as well as exact git and mercurial references. `haxelib dev` libs are ignored (with a warning), and try to resolve to "non-dev" version for the manifest file, allowing you to work with dev libs locally without altering the pin file.

I switched to file output instead of stdout for better cross-platform support, and API consistency.
"manifest" terminology comes from https://learn.microsoft.com/en-us/vcpkg/users/manifests

---
Original PR description below:

This PR adds a new _lock_ command which, as in other library managers, "locks" the current state of the libraries in the project. This is useful for reproducing the exact state of dependencies in a project (on a new machine or in Docker), especially when working with libraries installed from git.

The output is in HXML (`-L lib:version ...`) and it is dumped to stdout instead of a file.
The intended workflow is following:
1. do package management operations
2. lock the dependencies: `haxelib lock > haxelib-lock.hxml`
3. apply the lockfile in another environment: `haxelib install --skip-dependencies --always haxelib-lock.hxml`

I've created this for my personal usage and providing this patch hoping it could be useful to more developers. If inappropriate, feel free to add notes/close the PR.
Before merging, support for locking versions of Mercurial libs (installed using `haxelib hg ...`) should be added. I am not a Mercurial user myself, so help would be greatly appreciated.

### CODE REVIEW

File: src/haxelib/client/Main.hx
Comment (kLabz) [inline src/haxelib/client/Main.hx:866]:
```diff
-                       final version = scope.getVersion(library.name);
+                       final version = try scope.getVersion(library.name) catch (e) {
+                               Cli.printError('Error: ${e.message}');
+                               continue;
+                       };
```

Dev libs that don't have "non dev" version installed will make `haxelib lock' abort early if you don't catch errors there.

File: src/haxelib/client/Main.hx
Comment (kLabz) [inline src/haxelib/client/Main.hx:878]: I'm still not sure about generating `-L` instead of `--lib` because there are still people dealing with older versions. Might be just me, though..

File: src/haxelib/client/Args.hx
Comment (kLabz) [inline src/haxelib/client/Args.hx:260]: I think I'd move that command from `Basic` into `Development` category

File: src/haxelib/api/Vcs.hx
Comment (kLabz) [inline src/haxelib/api/Vcs.hx:359]: Well yeah it could.. but "most of the time it will work all the time" :shrug: 

Might need some error handling to recover just in case?

File: src/haxelib/api/Vcs.hx
Comment (kLabz) [inline src/haxelib/api/Vcs.hx:443]: Ok so I was able to get an implementation:
```haxe
		final oldCwd = Sys.getCwd();
		Sys.setCwd(libPath);

		final ret: VcsData = {
			url: run(["paths", "default"], true).out.trim(),
			ref: run(["identify", "-i", "--debug"], true).out.trim(),
		};

		Sys.setCwd(oldCwd);
		return ret;
```
**But** that doesn't work for installation, since `-L foo:hg:...#rev` will try to install `rev` as being a branch exclusively.

I had to update `clone` too:
```diff
        public function clone(libPath:String, url:String, ?branch:String, ?version:String, ?debugLog:(msg:String)->Void):Void {
                final vcsArgs = ["clone", url, libPath];
 
-               if (branch != null) {
+               if (branch != null && version != null) {
                        vcsArgs.push("--branch");
                        vcsArgs.push(branch);
-               }
 
-               if (version != null) {
                        vcsArgs.push("--rev");
                        vcsArgs.push(version);
+               } else if (branch != null) {
+                       vcsArgs.push("--updaterev");
+                       vcsArgs.push(branch);
+               } else if (version != null) {
+                       vcsArgs.push("--updaterev");
+                       vcsArgs.push(version);
                }
 
                if (run(vcsArgs, debugLog).code != 0)
                        throw VcsError.CantCloneRepo(this, url/*, ret.out*/);
        }
```

But I'd like to add proper tests for that. Only tested with `https://www.mercurial-scm.org/repo/hello` locally with:
```sh
haxelib hg test-tag https://repo.mercurial-scm.org/hello tip
haxelib hg test-branch https://repo.mercurial-scm.org/hello default
haxelib hg test-branch-tag https://repo.mercurial-scm.org/hello default "" tip
haxelib hg test-rev https://repo.mercurial-scm.org/hello 82e55d328c8ca4ee16520036c0aaace03a5beb65
```

Which all worked

**Comment by kLabz** [comment]
Thanks!
I think this can be merged once the above points have been handled

File: src/haxelib/client/Main.hx
Comment (j-jzk) [reply src/haxelib/client/Main.hx:878]: People using `haxelib lock` won't be using versions not supporting `-L` :)
I will change it to `--lib` anyway for clarity.

Comment (kLabz) [reply src/haxelib/client/Main.hx:878]: > People using haxelib lock won't be using versions not supporting -L :)
I disagree. I would use haxelib lock but I also support CI for multiple haxe versions.

**Comment by j-jzk** [comment]
Hopefully I will finish this till the end of the week.

**Comment by kLabz** [comment]
This PR is now ready as far as I'm concerned.
Will wait a bit before merging in case someone has some argument against it.

**Review by kLabz — APPROVED** [review]

File: src/haxelib/client/Args.hx
Comment (Simn) [inline src/haxelib/client/Args.hx:281]: This doesn't seem like the most relevant description...

File: src/haxelib/client/Main.hx
Comment (Simn) [inline src/haxelib/client/Main.hx:852]: wtf does "dumping lock from global" mean?

File: src/haxelib/client/Main.hx
Comment (Simn) [inline src/haxelib/client/Main.hx:863]: If this is an error, does it mean I cannot use the command at all as soon as I have a single dev library? That seems rather limiting... Shouldn't it rather warn about this and ignore dev libraries?

Comment (kLabz) [reply src/haxelib/client/Main.hx:863]: Ah, right I wanted to use "Warning" wording there.

I do want only warnings because I often have lock files that I want to keep working, but while actually working locally I also often work on some of these libs, so I use `haxelib dev` on them.

Current behavior warns about dev libs being present and so their line in output being potentially outdated, but still goes on.

File: src/haxelib/client/Args.hx
Comment (kLabz) [reply src/haxelib/client/Args.hx:281]: Is "output an hxml file with installation instructions for installed libraries" any better?

Comment (Simn) [reply src/haxelib/client/Args.hx]: Uh, barely...

I'd expect a lock command to mention, you know, locking? The fact that HXML files are involved doesn't even seem particularly important.

Comment (kLabz) [reply src/haxelib/client/Args.hx]: Well that's the thing, it doesn't lock anything. Maybe the issue is more with the command name?

Comment (Simn) [reply src/haxelib/client/Args.hx]: Are you saying we have a lock command that creates a lock file that doesn't lock anything? I'm not sure I understand what's happening here...

Comment (Simn) [reply src/haxelib/client/Args.hx]: Oh so that's what happens here... in this case I agree that "lock" is not a good name, and the description should probably end with "for use with `haxelib install`".

Comment (kLabz) [reply src/haxelib/client/Args.hx]: Hopefully this version is better (check updated PR description)

**Comment by j-jzk** [comment]
Thanks for all the work :) Just sat down to work on this and saw you'd already finished it (much better than I would have).

**Comment by kLabz** [comment]
Hope it still matches your workflow! Changed into a file arg so that windows people don't freak out :sweat_smile:

**Comment by kLabz** [comment]
@Simn is it ok to merge this?
Documentation will wait a bit (for a new haxelib release), and can be done on a separate PR.

**Pull request merged**
